### ISSUE TITLE
decreased performance on inventory loop

### ISSUE THREAD

**Issue by lassizci** [ansible/ansible#36614]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
- core

##### ANSIBLE VERSION

```
ansible 2.6.0 (devel a7db945ac6) last updated 2018/02/22 18:20:50 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/lpolonen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lpolonen/git/ansible/lib/ansible
  executable location = /home/lpolonen/git/ansible/bin/ansible
  python version = 2.7.14 (default, Dec 11 2017, 16:08:01) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]

```

```
ansible 2.5.0rc1 (detached HEAD e77c1b9e42) last updated 2018/02/22 18:51:31 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/lpolonen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/lpolonen/git/ansible/lib/ansible
  executable location = /home/lpolonen/git/ansible/bin/ansible
  python version = 2.7.14 (default, Dec 11 2017, 16:08:01) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

##### CONFIGURATION
default

##### OS / ENVIRONMENT
Fedora 26

##### SUMMARY
Close to 3x performance regression when looping through variables with debug.

##### STEPS TO REPRODUCE
inventory and variable setup:
```
echo "testvars:" >> testvars.yml
for i in $(seq -w 1 100); do echo "testhost-$i" >> inventory; echo "  testhost-${i}:" >> testvars.yml; echo "    test_var: ${i}" >> testvars.yml; done
```

playbook test_with_vars.yml:
```
- hosts: all
  connection: local
  gather_facts: no
  vars_files:
    - testvars.yml
  tasks:
    - debug:
        var: testvars[item].test_var
      with_items: "{{ groups.all }}"
```

to test:
```
time ansible-playbook -i inventory test_with_vars.yml 
```

##### EXPECTED RESULTS

Results should be on the same level
##### ACTUAL RESULTS

2.4.3.0
```
real	0m31,077s
user	1m43,971s
sys	0m8,338s
```
2.5.0rc1:
```
real	1m25,006s
user	4m55,067s
sys	0m12,774s
```

devel:
```
real	1m26,919s
user	5m6,903s
sys	0m13,164s
```

**Comment by lassizci** [comment]
originally tested with just `devel` but seems to affect 2.5.0 too..

**Comment by mkrizek** [comment]
cc @jimi-c @sivel

**Comment by sivel** [comment]
A git bisect indicates this started at bbd6b8bb42cc8fa48157ce5db0efb0a836c94917

In eca3fcd214903595db4d4e2f06813b74f1511e85: 0m3.499s
In bbd6b8bb42cc8fa48157ce5db0efb0a836c94917: 0m9.345s

This is with the reproducer provided, and using python3.  Py2 on I get 0m2.823s/0m13.801s respectively.

cc @bcoca

**Comment by bcoca** [comment]
weird, one of the few things i don't touch in that commit is vars loading

**Comment by sivel** [comment]
I agree, I was surprised it was that commit.  It is a big commit, and it could be hard to pin point. Between those commits I can visually see a degradation in performance between each debug.  I didn't profile, but it was easy to spot.

**Comment by bcoca** [comment]
I suspect this is not related to vars loading, as testing just that, shows no diff, this is probably a loop/display issue or even inventory looping.

using my play:
```
- hosts: localhost
  gather_facts: false
  vars_files:
    - testvars.yml
  tasks:
    - debug: var=testvars['testhost-007']['test_var']
```
10 on 'devel' runs, all mostly the same:
```
real    0m0.544s
user    0m0.466s
sys     0m0.067s
```

10 on '2.4' runs, all mostly the same:
```
real    0m0.713s
user    0m0.624s
sys     0m0.079s

```
so the vars loading is actually faster, but by very little.

**Comment by bcoca** [comment]
Added 'inventory looping' and 1 dynamic var access

```
- hosts: all
  gather_facts: false
  vars_files:
    - testvars.yml
  tasks:
    - debug: var=testvars[inventory_hostname]['test_var']
```

devel:
```
real    0m1.524s
user    0m4.565s
sys     0m1.077s
```

stable-2.4:
```
real    0m1.413s
user    0m2.838s
sys     0m0.935s
```

**Comment by bcoca** [comment]
Loop over hosts in tasks also:
```
- hosts: all
  gather_facts: no
  vars_files:
    - testvars.yml
  tasks:
    - debug:
        var: testvars[inventory_hostname].test_var
      with_items: "{{ groups.all }}"
```

devel
```
real    0m58.187s
user    4m52.886s
sys     0m6.699s
```

stable-2.4
```
real    0m36.839s
user    2m46.282s
sys     0m5.675s
```

one thing to note here, is that 2.4 is currently displaying extra info it should not during debug.

**Comment by bcoca** [comment]
looping over hosts and in task, but constant var access
```
- hosts: all
  gather_facts: no
  vars_files:
    - testvars.yml
  tasks:
    - debug:
        var: testvars['testhost-007'].test_var
      with_items: "{{ groups.all }}
```

devel
```
real    0m57.710s
user    4m49.082s
sys     0m6.949s
```

stable-2.4
```
real    0m36.970s
user    2m51.909s
sys     0m6.197s
```

**Comment by bcoca** [comment]
testing pure inventory loading `time ansible --list-hosts -i ./inventory all` ... gives me no change across versions.

so this seems 'looping related'.

**Comment by lassizci** [comment]
Yeah, perhaps not related to variable loading itself, as the results seem to be somewhat in line with my original tests (using vars_files to try to workaround slow hostvar access) with just slightly different playbook:

```
- hosts: all
  connection: local
  gather_facts: no
  vars_files:
    - testvars.yml
  tasks:
    - debug:
        var: hostvars[item].test_var
      with_items: "{{ groups.all }}"
```

That gives me (n= number of hosts in the inventory)
```
-f5, n=500, 2.4.3.0
real    17m40.776s
user    30m29.314s
sys     2m55.386s

-f5, n=100, 2.4.3.0
real    0m41.568s
user    1m10.218s
sys     0m7.064s


-f 5, n=500, ansible 2.6.0 (devel bd43776acf)
real    43m8.893s
user    78m56.036s
sys     4m9.149s

-f5, n=100, ansible 2.6.0 (devel bd43776acf)
real    1m38.128s
user    2m58.199s
sys     0m9.669s
```

Forks in this case doesn't seem to help beyond cpu cores. Going to 20 I usually hit:
```
ERROR! Unexpected Exception, this is probably a bug: [Errno 12] Cannot allocate memory
```
which probably depends on available memory.

**Comment by mkrizek** [comment]
With the original reproducer on the current devel I get:
```
real	0m2.031s
user	0m1.991s
sys	0m0.109s
```

`git bisect`'d a 'fix' to 0d0bdbec621e37e428dc88ed44ca2caaaf74d851.

**Comment by bcoca** [comment]
he, that is faster than in 2.4 ... going to close this then as per above.
